upstairs/src/lib.rs

```diff
@@ -8947,21 +9004,43 @@ pub struct Guest {
     /// it can be read from a `&self` reference.
     block_size: AtomicU64,
 
-    /// Backpressure is implemented as a delay on host operations
+    /// Backpressure is implemented as a delay on host write operations
     backpressure_us: AtomicU64,
 
     /// Backpressure configuration, as a starting point and max delay
     backpressure_config: BackpressureConfig,
+
+    /// Lock held during backpressure delay
+    ///
+    /// Without this lock, multiple tasks could submit jobs to the upstairs and
+    /// wait in parallel, which defeats the purpose of backpressure (since you
+    /// could send arbitrarily many jobs at high speed by sending them from
+    /// different tasks).
+    backpressure_lock: Mutex<()>,
 }
 
 /// Configuration for host-side backpressure
 ///
-/// Backpressure adds an artificial delay to host messages to maintain the
-/// downstairs queue at a particular length.
+/// Backpressure adds an artificial delay to host write messages (which are
+/// otherwise acked immediately, before actually being complete).  The delay is
+/// varied based on two metrics:
+///
+/// - number of write bytes outstanding
+/// - queue length as a fraction (where 1.0 is full)
+///
+/// These two metrics are used for quadratic backpressure, picking the larger of
+/// the two delays.
 #[derive(Copy, Clone, Debug)]
 struct BackpressureConfig {
-    start: f32,
-    max_delay: Duration,
+    /// When should backpressure start (in bytes)?
+    bytes_start: u64,
+    /// Scale for byte-based quadratic backpressure
+    bytes_scale: f64,
+
+    /// When should queue-based backpressure start?
+    queue_start: f64,
+    /// Maximum queue-based delay
```

Comment:
Will this queue based start consider IO types other than `Write`/`WriteUnwritten` when
deciding to apply backpressure?

I don't see backpressure checks in other IO types, but I'm wondering if there presence
in the work queue would trigger writes to slow down?

Reply:
Good catch – you're correct that if the queue has a bunch of non-write jobs, the queue-based backpressure will take them into account when delaying writes (and backpressure doesn't delay anything else).

This is a little weird, but I'm inclined to leave it as-is; the fundamental goal here is to avoid killing the upstairs due to hitting `MAX_ACTIVE_COUNT`, which counts every job in the queue.

(I'm also not sure how non-write jobs would manage to clog up the queue!)

Reply:
Yeah, I also think it's fine to leave here, just wanted to be sure I understood that
any IO on the queue can impact backpressure.